### PR TITLE
ENH: Add np.asanycontiguousarray

### PR DESCRIPTION
**Pull request by pitrou**
Numpy has np.ascontiguousarray() and np.asfortranarray(), but doesn't have
a function to return an array of either contiguity depending on the input,
making a copy only when necessary.

### CODE REVIEW

**Comment by jakirkham** [comment]
Probably should post it to the [NumPy mailing list]( http://www.scipy.org/scipylib/mailing-lists.html ) if you haven't already. If you have, maybe a link to the mailing list entry would be nice and vice versa.

File: numpy/core/numeric.py [hunk 1]
```diff
@@ -604,6 +607,53 @@ def asfortranarray(a, dtype=None):
...
+    """
+    try:
+        is_f_contiguous = a.flags.f_contiguous
+    except AttributeError:
+        # Not a ndarray
+        is_f_contiguous = False
+    if is_f_contiguous:
+        return array(a, dtype, copy=False, order='F', ndmin=1)
+    else:
+        return array(a, dtype, copy=False, order='C', ndmin=1)
```
Comment (seberg) [inline numpy/core/numeric.py]: Not sure, I like the idea too much, but that is not a deal breaker (also the `any` has a bit of a name clash with `asanyarray`). But I think you can simplify this to a single line function:
```
return array(a, dtype, copy=False, order="A", ndmin=1)
```
(not that I like the `ndmin=1` either, what is so bad about 0-D arrays? But I guess the ascontiguousarray functions do that too). The whole f_contiguous stuff should be handled by "A" is my guess.

**Comment by charris** [comment]
Needs release note.

**Comment by njsmith** [comment]
If @seberg is right that this is just `copy=False, order="A"`, then I'm -0.5. No need to clutter the top level namespace with unique aliases for every combination of arguments that can passed to `array`. (This also avoids the sticky question of whether people actually want `ndmin=1`.)

**Comment by charris** [comment]
I'm also -0.5 for this. We need to limit the growth of small functions at some point.

**Comment by seberg** [comment]
Unfortunatly, I am not right. "A" typically works like that, so I
assumed, but it seems for the no-copy case, it is happy to allow non
-contiguous arrays here.

**Comment by njsmith** [comment]
Can we fix A?

**Comment by seberg** [comment]
If it is not really used much, I guess it is an option. It is
documented as not caring about contiguity at all for non-copy mode. But
there seems little to no reason in giving "A" when None/not giving
anything will do the same thing.

**Comment by pitrou** [comment]
Well, "A" typically means "any layout", so I'm not sure why changing it to "any contiguous layout" in a single API would make things easier for the user.

**Comment by pitrou** [comment]
(I certainly wouldn't guess that calling `np.array` with order "A" would ensure a contiguous array is returned, myself :-))

**Comment by njsmith** [comment]
Another option would be to add another order value, like `order="C-or-F"` or `order="either"`.

**Comment by seberg** [comment]
None of the options sound very convincing to me on first sight :(. "A" stands for any, that is true, but it seems (almost?) only used in the "fortran if not C (and actually fortran)" sense, though sometimes in terms of iteration order and not memory layout (we have "K" as a weaker "A"). "ForC" or variations of this, may make some sense since we use it in `arr.flags.forc`. It might also make sense for that wonky "require" function, which I could not remember though I looked it up 2 minutes ago ;).

**Comment by charris** [comment]
This is looking like a reasonable addition. Does anyone object? @pitrou Needs mention in 1.12.0 release notes.

**Comment by njsmith** [comment]
I'm still -0 I guess since it isn't clear to me how this could have broad utility for multiple users (what is the use case, BTW?), and because it isn't clear that adding a new function is better than adding a new option to `np.array` (which would be more orthogonal and reduce doc/namespace clutter). But only -0 :-)

**Comment by pitrou** [comment]
The use case we have in mind is writing wrappers for third-party libraries such as BLAS, which often accept either C- or F-ordered arrays, but not non-contiguous arrays.

**Comment by efiring** [comment]
I think the name conflict with `asanyarray` is a serious drawback.  The only way to keep the name and solve the clash would be to include the `subok=True` kwarg.

**Comment by efiring** [comment]
A better way to add the desired functionality would be via adding a kwarg to `ascontiguousarray`.

**Comment by seberg** [comment]
Good point, giving that an order keyword argument should be pretty uncontroversial. @pitrou maybe that would be a solution to continue here?

**Comment by charris** [comment]
@pitrou I'm going to close this. @efiring's suggestion looks like a good one.

**Pull request closed without merging**